File: Algorithm/Python/413. Arithmetic Slices.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-10-15 17:16:39
# @Last modified by:   WuLC
# @Last Modified time: 2016-10-15 17:22:44
# @Email: liangchaowu5@gmail.com

# Math, O(n) time, O(1) space
# find the array of difference of all consecutive numbers, 
# and go through the difference array to find those continious same number and count the number of arithmetic slices
class Solution(object):
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        diff = [A[i+1]-A[i] for i in xrange(len(A)-1)]
        pre, count, result = None, 0, 0
        for i in xrange(len(diff)):
            if pre == None or diff[i] == pre:
                count += 1
            else:
                if count > 1:
                    count -= 1
                    result += (1+count)*count/2
                count = 1
            pre = diff[i]
        if count > 1:
            count -= 1
            result += (1+count)*count/2
        return result

File: Algorithm/Java/413. Arithmetic Slices.java
/*
* @Author: WuLC
* @Date:   2017-06-15 16:59:15
* @Last Modified by:   WuLC
* @Last Modified time: 2017-06-15 17:01:55
* @Email: liangchaowu5@gmail.com
*/

// O(n) time, O(1) space
// find the longest arithmetic slice each time, then calculate all possible sub slices within this slice
public class Solution 
{
    public int numberOfArithmeticSlices(int[] A) 
    {
        int start = 0 , end = 1, result = 0;
        while (end < A.length - 1)
        {
            if (A[end+1] - A[end] == A[end] - A[end-1]) end++;
            else
            {
                if(end - start >= 2) result += ((end - start)*(end - start - 1)>>1);
                start = end;
                end++;
            }
        }
        if (end == A.length - 1 && end - start >= 2) result += ((end - start)*(end - start - 1)>>1);
        return result;
    }
}